graphql_compiler/interpreter/typedefs.py

```diff
@@ -0,0 +1,208 @@
+    - Interpreters perform lazy evaluation by default: if exactly 3 query results are requested,
+      then only the minimal data necessary to produce *exactly 3 rows' worth* of outputs is loaded.
+    - When computing a particular result, data loading for output fields is deferred
+      until *after* all filtering operations have been completed, to minimize data loads.
+    - Data caching is easy to implement within this API -- simply have
+      your API function's implementation consult a cache before performing the requested operation.
+    - Batch-loading of data can be performed by simply advancing the input generator multiple times,
+      then operating on an entire batch of input data before producing corresponding outputs:
+
+        def project_property(
+            self,
+            data_contexts: Iterable[DataContext[DataToken]],
+            current_type_name: str,
+            field_name: str,
+            **hints: Any
+        ) -> Iterable[Tuple[DataContext[DataToken], Any]]:
+            for data_context_batch in funcy.chunks(30, data_contexts):
+                # Data for 30 entries is now in data_context_batch, operate on it in bulk.
+                results_batch = compute_results_for_batch(
+                    data_context_batch, current_type_name, field_name
+                )
+                yield from results_batch
+
+    Additionally, each of the four methods in the API takes several kwargs whose names
+    end with the suffix "_hints", in addition to the catch-all "**hints: Any" argument. These
+    provide each function with information about how the data it is currently processing will
+    be used in subsequent operations, and can therefore enable additional interesting optimizations.
+    Use of these hints is totally optional (the library always assumes that the hints weren't used),
+    so subclasses of InterpreterAdapter may even safely ignore these kwargs entirely -- for example,
+    if the "runtime_arg_hints" kwarg is omitted in the method definition, at call time its value
+    will go into the catch-all "**hints" argument instead.
+
+    The set of hints (and the information each hint provides) could grow in the future. Currently,
+    the following hints are offered:
+    - runtime_arg_hints: the values of any runtime arguments provided to the query for use in
+      filtering operations (recall queries' "$foo" filter parameter syntax).
+    - used_property_hints: the property names within the scope relevant to the called function that
+      the query will eventually need, e.g. for filtering on, or to output as the final result.
+    - filter_hints: information about the filters applied within the scope relevant to the called
+      function, such as "which filtering operation is being performed?" and "with which arguments?"
+    - neighbor_hints: information about the edges originating from the scope relevant to the called
+      function that the query will eventually need to expand.
```

Comment:
What does "the scope relevant to the called function" mean?  Also, is the scope relevant to the called function, or to the particular invocation of the function?

Reply:
I rewrote this part to explicitly define which scope is in play for each function, let me know what you think.